File: worker/src/telegram.ts
const TELEGRAM_LIMIT = 4096;

export function escapeHtml(text: string): string {
  return text
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;");
}

function splitMessage(text: string, limit = TELEGRAM_LIMIT): string[] {
  if (text.length <= limit) return [text];
  const chunks: string[] = [];
  let current = "";
  for (const para of text.split("\n\n")) {
    const candidate = current ? `${current}\n\n${para}` : para;
    if (candidate.length > limit && current) {
      chunks.push(current);
      current = para;
    } else {
      current = candidate;
    }
  }
  if (current) chunks.push(current);
  // A single paragraph longer than the limit is still possible in theory.
  return chunks.flatMap((c) => {
    const parts: string[] = [];
    for (let i = 0; i < c.length; i += limit) parts.push(c.slice(i, i + limit));
    return parts;
  });
}

export async function sendMessage(token: string, chatId: string | number, html: string): Promise<void> {
  for (const chunk of splitMessage(html)) {
    const resp = await fetch(`https://api.telegram.org/bot${token}/sendMessage`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        chat_id: chatId,
        text: chunk,
        parse_mode: "HTML",
        disable_web_page_preview: true,
      }),
    });
    const payload = (await resp.json().catch(() => ({}))) as { ok?: boolean; description?: string };
    if (!payload.ok) {
      throw new Error(`sendMessage failed (${resp.status}): ${payload.description ?? "unknown"}`);
    }
  }
}

export async function sendTyping(token: string, chatId: string | number): Promise<void> {
  await fetch(`https://api.telegram.org/bot${token}/sendChatAction`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ chat_id: chatId, action: "typing" }),
  }).catch(() => {});
}

File: src/tech_news_summarizer/telegram.py
"""Format parsed issues as Telegram HTML and send them via the Bot API."""

import datetime as dt
import html

import requests

from .config import CATEGORY_NAMES
from .fetcher import Issue

TELEGRAM_LIMIT = 4096


class TelegramError(Exception):
    pass


def format_digest(category: str, date: dt.date, issue: Issue) -> str:
    """Per-newsletter digest (fallback path when AI merging is unavailable)."""
    name = CATEGORY_NAMES.get(category, category)
    lines = [f"<b>📰 TLDR {html.escape(name)} — {date.isoformat()}</b>"]
    if issue.tagline:
        lines.append(f"<i>{html.escape(issue.tagline)}</i>")
    lines.append("")
    lines.extend(_format_issue_body(issue))
    return "\n".join(lines).strip()


def format_combined_digest(date: dt.date, categories: list[str], issue: Issue) -> str:
    """Single merged digest across all fetched newsletters."""
    names = ", ".join(CATEGORY_NAMES.get(c, c) for c in categories)
    lines = [
        f"<b>📰 TLDR Daily Digest — {date.isoformat()}</b>",
        f"<i>{html.escape(names)}</i>",
        "",
    ]
    lines.extend(_format_issue_body(issue))
    return "\n".join(lines).strip()


def _format_issue_body(issue: Issue) -> list[str]:
    lines: list[str] = []
    for section in issue.sections:
        if section.name:
            lines.append(f"<b><u>{html.escape(section.name)}</u></b>")
        for story in section.stories:
            headline = html.escape(story.headline)
            if story.url:
                lines.append(
                    f'• <a href="{html.escape(story.url, quote=True)}"><b>{headline}</b></a>'
                )
            else:
                lines.append(f"• <b>{headline}</b>")
            lines.append(f"  {html.escape(story.blurb)}")
            lines.append("")
    return lines


def split_message(text: str, limit: int = TELEGRAM_LIMIT) -> list[str]:
    if len(text) <= limit:
        return [text]
    chunks: list[str] = []
    current = ""
    for para in text.split("\n\n"):
        candidate = f"{current}\n\n{para}" if current else para
        if len(candidate) > limit and current:
            chunks.append(current)
            current = para
        else:
            current = candidate
    if current:
        chunks.append(current)
    # A single paragraph longer than the limit is still possible in theory.
    return [c[i : i + limit] for c in chunks for i in range(0, len(c), limit)]


def send_message(token: str, chat_id: str, html_text: str) -> None:
    for chunk in split_message(html_text):
        resp = requests.post(
            f"https://api.telegram.org/bot{token}/sendMessage",
            json={
                "chat_id": chat_id,
                "text": chunk,
                "parse_mode": "HTML",
                "disable_web_page_preview": True,
            },
            timeout=30,
        )
        payload = resp.json() if resp.headers.get("content-type", "").startswith("application/json") else {}
        if not payload.get("ok"):
            raise TelegramError(f"sendMessage failed ({resp.status_code}): {resp.text[:300]}")
